aptos-move/block-executor/src/executor.rs

```diff
@@ -636,103 +633,32 @@ where
                 }
             }
         }
-        (patched_resource_write_set, write_set_keys)
+        patched_resource_write_set
     }
 
     // Parse the input `value` and replace delayed field identifiers with
     // corresponding values
     fn replace_ids_with_values(
-        value: Arc<T::Value>,
+        value: &T::Value,
         layout: &MoveTypeLayout,
         latest_view: &LatestView<T, S, X>,
-        delayed_field_keys: &HashSet<T::Identifier>,
-    ) -> Option<T::Value> {
-        if let Some(value_bytes) = value.bytes() {
-            match latest_view.replace_identifiers_with_values(value_bytes, layout) {
-                Ok((patched_bytes, delayed_field_keys_in_resource)) => {
-                    if !delayed_field_keys.is_disjoint(&delayed_field_keys_in_resource) {
-                        let mut patched_value = value.as_ref().clone();
-                        patched_value.set_bytes(patched_bytes);
-                        Some(patched_value)
-                    } else {
-                        None
-                    }
-                },
-                Err(_) => unreachable!("Failed to replace identifiers with values in read set"),
+    ) -> T::Value {
+        if let Some(mut value) = value.from_read_to_modification() {
+            if let Some(value_bytes) = value.bytes() {
+                let (patched_bytes, _) = latest_view
+                    .replace_identifiers_with_values(value_bytes, layout)
+                    .unwrap();
+                value.set_bytes(patched_bytes);
+                value
+            } else {
+                unreachable!("Value to be exchanged doesn't have bytes: {:?}", value)
```

Comment:
Can we have a zero byte change?

Reply:
Zero bytes cannot have an identifier we need to exchange